arangod/Cluster/ClusterInfo.cpp

```diff
@@ -1092,7 +1119,40 @@ void ClusterInfo::loadPlan() {
     // we can safely keep this iterator around because we hold the read-lock on _planProt here.
     // reusing the lookup positions helps avoiding redundant lookups into _plannedCollections
     // for the same database
-    AllCollections::const_iterator existingCollections = _plannedCollections.find(databaseName);
+
+    AllCollections::const_iterator existingCollections,
+      stillExistingCollections = newCollections.find(databaseName);
+    {
+      READ_LOCKER(guard, _planProt.lock);
+      existingCollections = _plannedCollections.find(databaseName);
+    }
+
+    if (stillExistingCollections != newCollections.end()) {
+      auto const& np = newPlan.find(databaseName);
+      if (np != newPlan.end()) {
+        auto nps = np->second->slice()[0];
+        for (auto const& ec : *(stillExistingCollections->second)) {
+          auto const& cid = ec.first;
+          if (!std::isdigit(cid.front())) {
+            continue;
+          }
+          collectionsPath.push_back(cid);
+          if (!nps.hasKey(collectionsPath))  { // collection gone
+            collectionsPath.push_back("shards");
+            READ_LOCKER(guard, _planProt.lock);
+            for (auto const& sh :
+                   VPackObjectIterator(_plan.find(databaseName)->second->slice()[0].get(collectionsPath))) {
+              auto const& shardId = sh.key.copyString();
+              newShards.erase(shardId);
+              newShardServers.erase(shardId);
+              newShardToName.erase(shardId);
+            }
+            collectionsPath.pop_back();
+          }
+          collectionsPath.pop_back();
```

Comment:
If the key is present in `nps`, it is not necessary to push and pop from collectionsPath.
```suggestion
          if (!nps.hasKey(collectionsPath))  { // collection gone
            collectionsPath.push_back(cid);
            collectionsPath.push_back("shards");
            READ_LOCKER(guard, _planProt.lock);
            for (auto const& sh :
                   VPackObjectIterator(_plan.find(databaseName)->second->slice()[0].get(collectionsPath))) {
              auto const& shardId = sh.key.copyString();
              newShards.erase(shardId);
              newShardServers.erase(shardId);
              newShardToName.erase(shardId);
            }
            collectionsPath.pop_back();
            collectionsPath.pop_back();
          }
```

Reply:
it is necessary.

Reply:
Ah, true.